pkg/orm/orm.go

```diff
@@ -442,105 +479,147 @@ func (o *orm) QueryTable(ptrStructOrTableName interface{}) (qs QuerySeter) {
+}
+
+// return sql.DBStats for current database
+func (o *ormBase) DBStats() *sql.DBStats {
+	if o.alias != nil && o.alias.DB != nil {
+		stats := o.alias.DB.DB.Stats()
+		return &stats
 	}
 	return nil
 }
 
-// begin transaction
-func (o *orm) Begin() error {
-	return o.BeginTx(context.Background(), nil)
+type orm struct {
+	ormBase
 }
 
-func (o *orm) BeginTx(ctx context.Context, opts *sql.TxOptions) error {
-	if o.isTx {
-		return ErrTxHasBegan
-	}
-	var tx *sql.Tx
+var _ Ormer = new(orm)
+
+func (o *orm) Begin() (TxOrmer, error) {
+	return o.BeginWithCtx(context.Background())
+}
+
+func (o *orm) BeginWithCtx(ctx context.Context) (TxOrmer, error) {
+	return o.BeginWithCtxAndOpts(ctx, nil)
+}
+
+func (o *orm) BeginWithOpts(opts *sql.TxOptions) (TxOrmer, error) {
+	return o.BeginWithCtxAndOpts(context.Background(), opts)
+}
+
+func (o *orm) BeginWithCtxAndOpts(ctx context.Context, opts *sql.TxOptions) (TxOrmer, error) {
 	tx, err := o.db.(txer).BeginTx(ctx, opts)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	o.isTx = true
-	if Debug {
-		o.db.(*dbQueryLog).SetDB(tx)
-	} else {
-		o.db = tx
+
+	_txOrm := &txOrm{
+		ormBase: ormBase{
+			alias: o.alias,
+			db:    &TxDB{tx: tx},
+		},
+		isClosed: false,
 	}
-	return nil
+
+	var taskTxOrm TxOrmer = _txOrm
+	return taskTxOrm, nil
 }
 
-// commit transaction
-func (o *orm) Commit() error {
-	if !o.isTx {
-		return ErrTxDone
-	}
-	err := o.db.(txEnder).Commit()
-	if err == nil {
-		o.isTx = false
-		o.Using(o.alias.Name)
-	} else if err == sql.ErrTxDone {
-		return ErrTxDone
-	}
-	return err
+func (o *orm) DoTx(task func(txOrm TxOrmer) error) error {
+	return o.DoTxWithCtx(context.Background(), task)
 }
 
-// rollback transaction
-func (o *orm) Rollback() error {
-	if !o.isTx {
-		return ErrTxDone
-	}
-	err := o.db.(txEnder).Rollback()
-	if err == nil {
-		o.isTx = false
-		o.Using(o.alias.Name)
-	} else if err == sql.ErrTxDone {
-		return ErrTxDone
+func (o *orm) DoTxWithCtx(ctx context.Context, task func(txOrm TxOrmer) error) error {
+	return o.DoTxWithCtxAndOpts(ctx, nil, task)
+}
+
+func (o *orm) DoTxWithOpts(opts *sql.TxOptions, task func(txOrm TxOrmer) error) error {
+	return o.DoTxWithCtxAndOpts(context.Background(), opts, task)
+}
+
+func (o *orm) DoTxWithCtxAndOpts(ctx context.Context, opts *sql.TxOptions, task func(txOrm TxOrmer) error) error {
+	_txOrm, err := o.BeginWithCtxAndOpts(ctx, opts)
+	if err != nil {
+		return err
 	}
+	panicked := true
+	defer func() {
+		if panicked || err != nil {
+			e := _txOrm.Rollback()
+			logs.Error("rollback transaction failed: %v", e)
+		} else {
+			e := _txOrm.Commit()
+			logs.Error("commit transaction failed: %v", e)
+		}
+	}()
+
+	var taskTxOrm = _txOrm
+	err = task(taskTxOrm)
+	panicked = false
 	return err
 }
 
-// return a raw query seter for raw sql string.
-func (o *orm) Raw(query string, args ...interface{}) RawSeter {
-	return newRawSet(o, query, args)
+type txOrm struct {
+	ormBase
+	isClosed   bool
```

Comment:
If you want to check `isClosed`, it means all invocation you should check, including `Read`, `ReadForUpdate`... If you don't check those invocations, you don't need to check `Commit` and `Rollback` here.

Reply:
Actually I think that users should be responsible for keeping the txOrm is thread-safe. We don't need to do that.